pjsip/src/pjsua-lib/pjsua_acc.c

```diff
@@ -2162,8 +2163,21 @@ static void update_keep_alive(pjsua_acc *acc, pj_bool_t start,
     if (start) {
 	pj_time_val delay;
 	pj_status_t status;
+	pjsip_generic_string_hdr *hsr = NULL;
+
+	static const pj_str_t STR_FLOW_TIMER  = { "Flow-Timer", 10 };
+
+	hsr = (pjsip_generic_string_hdr*)
+	      pjsip_msg_find_hdr_by_name(param->rdata->msg_info.msg,
+					 &STR_FLOW_TIMER, hsr);
+
+	if (hsr) {
+	    unsigned tmr_exp = (pj_rand() % 20) + 80;
+	    acc->rfc5626_flowtmr = pj_strtoul(&hsr->hvalue) * tmr_exp / 100;
+	}
```

Comment:
From the RFC:
```
If the
   UA uses the server-recommended keep-alive frequency it SHOULD send
   its keep-alives so that the interval between each keep-alive is
   randomly distributed between 80% and 100% of the server-provided
   time.  For example, if the server suggests 120 seconds, the UA would
   send each keep-alive with a different frequency between 95 and 120
   seconds.
```

The randomisation applies when sending each keep alive, so it shouldn't be here, but when scheduling the timer below.

This is probably not very major, but also note that the random number selection should be between a lower bound and an upper bound, instead of percentage based (percentage based calculation may cause certain numbers to never be picked).
```
The UA MUST select a random number between a fixed or
   configurable upper bound and a lower bound, where the lower bound is
   20% less then the upper bound.
```

Reply:
The random number recommendation is when `Flow-timer` is not provided on the response.
```
When a Flow-Timer header field is not provided in the most recent
   success registration response, the proper selection of keep-alive
   frequency is primarily a trade-off between battery usage and
   availability.  The UA MUST select a random number between a fixed or
   configurable upper bound and a lower bound, where the lower bound is
   20% less then the upper bound.
```
In this case we already have ```ka_interval``` settings, and it will change the current timer calculation.

Reply:
True, but I think the spirit is the same.
```If the
   UA uses the server-recommended keep-alive frequency it SHOULD send
   its keep-alives so that the interval between each keep-alive is
   randomly distributed between 80% and 100% of the server-provided
   time.  For example, if the server suggests 120 seconds, the UA would
   send each keep-alive with a different frequency between 95 and 120
   seconds.
```
Now let's say the server suggests 1000 seconds (the RFC recommends 840 sec when battery power is a concern in another section, and here I picked a nice round number for easier calculation). Now the interval should be randomly distributed between 800/799 to 1000 seconds. With percentage based calculation, it would be 810, 820, 830, ..., and so on, and will never pick 801, 802, ..., 811, 812, etc. In other words, there will be numbers that have no distribution at all.

Again, this is a minor issue and probably won't matter much in practice, but since the calculation is as simple, I believe we should adopt this instead?

Note: the current percentage-based approach is actually not wrong, but it needs to be floating point in order to work, i.e. it needs to choose a random floating point number between 80% to 100%.

Reply:
I was saying that the percentage approach is actually not wrong, but still, I won't recommend it since it makes the calculation much more complicated.

Instead of choosing an (integer) random number between 80 to 100, why not choose an integer between lower and upper bound as the RFC suggested? The calculation is the same and as simple, i.e.:
```
unsigned lower_bound = (unsigned)timer * 0.8f;
delay = pj_rand() % (timer - lower_bound) + lower_bound;
```